### ISSUE TITLE
hashindex: malloc buckets failed (Cannot allocate memory)

### ISSUE THREAD

**Issue by No15pecial** [borgbackup/borg#4765]
## Have you checked borgbackup docs, FAQ, and open Github issues?

I have tried. 

## Is this a BUG / ISSUE report or a QUESTION?

Question

## System information. For client/server mode post info for both machines.
Linux: Ubuntu 18.04 bionic 
Dell Optiplex 755 dual core 4 GB memory (I know it's low) 

#### Your borg version (borg -V).
Borg: 1.1.5  Python: CPython 3.6.8

#### Operating system (distribution) and version.
Linux: Ubuntu 18.04 bionic 

#### Hardware / network configuration, and filesystems used.
Dell Optiplex 755 dual core 4 GB memory (I know it's low). 1 GB network, Exported R/Root NFS from Network Appliance NAS backed up to Exported R/W NFS Synology NAS 

#### How much data is handled by borg?
About 2.3 TB (seems to work the first few runs then it starts hanging up).

## Describe the problem you're observing.

```
eval   borg create -verbose  --stats --chunker-params 19,23,21,4095 \
  --show-rc --progress --exclude-caches \
  /Synology-Mounts/BorgHoard/GDCBorgRepo::BorgBK1-2019-09-24 \
  /NewCorpFiler-Mounts/Vol1/ \
  /NewCorpFiler-Mounts/GDC_IT/GDC-IT \
  /NewCorpFiler-Mounts/GDC_ENG \
  --exclude '*/.snapshot' --show-rc

hashindex: malloc buckets failed(Cannot allocate memory)r-M...PAGES/QPR-20.0.doc
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4253, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4185, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 152, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 547, in do_create
    create_inner(archive, cache)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 511, in create_inner
    read_special=args.read_special, dry_run=dry_run, st=st)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 621, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 621, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 621, in _process
    read_special=read_special, dry_run=dry_run)
  [Previous line repeated 4 more times]
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 598, in _process
    status = archive.process_file(path, st, cache)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 1033, in process_file
    return status
  File "/usr/lib/python3.6/contextlib.py", line 88, in __exit__
    next(self.gen)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 887, in create_helper
    self.add_item(item)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 444, in add_item
    self.items_buffer.add(item)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 239, in add
    self.flush()
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 256, in flush
    self.chunks.append(self.write_chunk(chunk))
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 272, in write_chunk
    id_, _, _ = self.cache.add_chunk(self.key.id_hash(chunk), chunk, self.stats, wait=False)
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 898, in add_chunk
    self.repository.put(id, data, wait=wait)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1076, in put
    self.index[id] = segment, offset
  File "src/borg/hashindex.pyx", line 214, in borg.hashindex.NSIndex.__setitem__
Exception: hashindex_set failed

Platform: Linux BorgBK1 4.15.0-45-generic #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019 x86_64
Linux: Ubuntu 18.04 bionic
Borg: 1.1.5  Python: CPython 3.6.8
PID: 2391  CWD: /root
sys.argv: ['/usr/bin/borg', 'create', '-verbose', '--stats', '--chunker-params', '19,23,21,4095', '--show-rc', '--progress', '--exclude-caches', '/Synology-Mounts/BorgHoard/GDCBorgRepo::BorgBK1-2019-09-24', '/NewCorpFiler-Mounts/Vol1/', '/NewCorpFiler-Mounts/GDC_IT/GDC-IT', '/NewCorpFiler-Mounts/GDC_ENG', '--exclude', '*/.snapshot', '--show-rc']
SSH_ORIGINAL_COMMAND: None

terminating with error status, rc 2
```

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Just about every night - again it seems to work for 2 or 3 days then it dies after that

**Comment by infectormp** [comment]
Try to reproduce this error with recent version of borg

**Comment by ThomasWaldmann** [comment]
Some questions / comments:
- what is the point of using `eval` here?
- you are having option arguments on both sides of positional arguments. this may not always work as you think, see "order matters" in our docs.

You obviously have a problem with borg trying to use more than available memory, so:
- how much swap is active and how much total memory is used short before it crashes?
- you could maybe add some more physical memory
- or reduce the memory needs of other stuff running in parallel with borg.
- do you see anything special in `atop` output while borg is running / short before it is crashing?
- if you have lots of files (high file count), the files cache needs quite some memory. you could maybe partition the source data set into multiple subsets and use 1 repo per subset.
- similar approach for high data volume and the memory needed by the chunks index.
- the files cache can be disabled, which would save quite some memory, but in that case borg would always read/chunk all source files, being a lot slower.

**Comment by ThomasWaldmann** [comment]
It might not fix this problem, but quite some others if you use a more recent borg version from the ubuntu borg maintainer's ppa. If you add that to your sources.list.d/, you'll get updates, too.

**Comment by No15pecial** [comment]
I do it in a cornjob via bash script, using variables to create an array of what to backup based on what gets mounted for that day.  The eval makes it so the variables get interpreted correctly.  I'll fix the -verbose thanks for pointing that out.  Keep in mind if I delete the REPO, recreate it and run this script it does run for a couple of days. After it gets to 2.3 TB (what I believe is all my data) it starts to error out rather quickly with the hashindex error. 

The actual script line is 
'''
 eval   borg create -verbose  --stats --chunker-params 19,23,21,4095 --show-rc --progress --exclude-caches  $BorgArchive  ${BKT[@]} --exclude '*/.snapshot' --show-rc
'''
https://www.unix.com/man-page/posix/1posix/eval/

The eval statement tells the shell to take eval’s arguments as command and run them through the command-line. It is useful in a situation like below:

In your script if you are defining a command into a variable and later on you want to use that command then you should use eval:

/home/user1 > a="ls | more"
/home/user1 > $a
bash: command not found: ls | more

The Above command didn't work as ls tried to list file with name pipe (|) and more. But these files are not there
Use eval and it will work  

/home/user1 > eval $a

**Comment by No15pecial** [comment]
There's a new version - I though I had the newest version   OK I'll look up how to do that 

P.S. Does replying to this email update this GitHub

**Comment by No15pecial** [comment]
atop borg not running 

```
ATOP - BorgBK1 2019/09/25  09:55:52 - 23m54s elapsed             2019/09/25  09:56:42                                                               --                                                     10
PRC |  sys    0.02s |  user   0.00s |               |               |  #proc    113 |  #trun      1 |  #tslpi    72  | #tslpu     0  | #zombie    0  | clones     0  |               |               | #exit      0  |
CPU |  sys       0% |  user      0% |  irq       0% |               |               |  idle    200% |  wait      0%  |               | steal     0%  | guest     0%  |               | curf 1.26GHz  | curscal   ?%  |
cpu |  sys       0% |  user      0% |  irq       0% |               |               |  idle    100% |  cpu001 w  0%  |               | steal     0%  | guest     0%  |               | curf 1.23GHz  | curscal   ?%  |
CPL |  avg1    0.01 |               |  avg5    0.01 |  avg15   0.00 |               |               |  csw      946  |               | intr     843  |               |               | numcpu     2  |               |
MEM |  tot     3.7G |  free    3.4G |  cache 149.2M |  dirty   0.1M |  buff   16.1M |  slab   59.8M |  slrec  28.9M  | shmem   0.9M  | shrss   0.0M  | shswp   0.0M  | vmbal   0.0M  | hptot   0.0M  | hpuse   0.0M  |
SWP |  tot   980.0M |  free  980.0M |               |               |               |               |                |               |               |               | vmcom 204.4M  |               | vmlim   2.8G  |
LVM |  BK1--vg-root |  busy      0% |               |  read       0 |  write      5 |  KiB/r      0 |  KiB/w      4  |               | MBr/s    0.0  | MBw/s    0.0  | avq     1.00  |               | avio 2.40 ms  |
DSK |           sda |  busy      0% |               |  read       0 |  write      3 |  KiB/r      0 |  KiB/w      6  |               | MBr/s    0.0  | MBw/s    0.0  | avq     1.00  |               | avio 4.00 ms  |
NET |  transport    |  tcpi       4 |  tcpo      20 |  udpi       0 |  udpo 0 |  tcpao KiB/r      0 |  tcppo      0  |               | tcprs      0  | tcpie      0  | tcpor      0  | udpnp      0  | udpie      0  |
NET |  network      |  ipi 49 |   5    read         |  ipo 78 |  20frw      0 |  deliv     49 |   5              0K  | dwrit     0K  |               |               | icmpi      0  |               | icmpo      0  |
NET |  enp0s25   0% |  pcki     164 |  pcko      22 |  sp 1000 Mbps |  si    9 Kbps |  so    3 Kbps |                | coll       0  | mlti       5  | erri       0  | erro       0  | drpi      81  | drpo       0  |

  PID SYSCPU USRCPU VGROW RGROW RDDSK WRDSK RUID EUID ST EXC THR S CPUNR CPU CMD 1/1|  deliv   5100 |                |               |               |               | icmpi    176  |               | icmpo     14  |
 1787 0.01s 0.01s211260K 928K 00 0K rootc    0 2404 |  sp   00 Mbps |  si   14 Kbps |  so    2 Kbps |              root            -- - 1ti     812  | erri          R            1   0% atop
    1         0.00s         0.00s            0K             0K            0K            0K        root             root            --          -           1         S            0          0%         systemd
 1589         0.00s         0.00s            0K             0K            0K            0K        administ         administ        --          -           1         S            0          0%         sshd
  349         0.00s         0.00s            0K             0K            0K           12K        root             root            --          -           1         S            0          0%         jbd2/dm-0-8
```

I think swap is small  

root@BorgBK1:~# swapon
NAME      TYPE      SIZE USED PRIO
/dev/dm-1 partition 980M   0B   -2

**Comment by No15pecial** [comment]
atop with Borg Running 

```
ATOP - BorgBK1 2019/09/25  10:08:57 - 36m59s elapsed             2019/09/25  10:10:07                                                               --                                                     10
PRC |  sys    0.36s |  user   0.93s |               |               |  #proc    121 |  #trun      1 |  #tslpi    77  | #tslpu     1  | #zombie    0  | clones     0  |               |               | #exit      0  |
CPU |  sys       3% |  user      9% |  irq       1% |               |               |  idle    107% |  wait     80%  |               | steal     0%  | guest     0%  |               | curf 1.37GHz  | curscal   ?%  |
cpu |  sys       3% |  user      9% |  irq       0% |               |               |  idle      8% |  cpu000 w 81%  |               | steal     0%  | guest     0%  |               | curf 1.52GHz  | curscal   ?%  |
cpu |  sys       1% |  user      0% |  irq       0% |               |               |  idle     99% |  cpu001 w  0%  |               | steal     0%  | guest     0%  |               | curf 1.21GHz  | curscal   ?%  |
CPL |  avg1    1.03 |               |  avg5    1.00 |  avg15   0.54 |               |               |  csw    25058  |               | intr   14613  |               |               | numcpu     2  |               |
MEM |  tot     3.7G |  free  104.8M |  cache  17.9M |  dirty   0.0M |  buff    1.6M |  slab   84.9M |  slrec  51.4M  | shmem   0.0M  | shrss   0.0M  | shswp   0.0M  | vmbal   0.0M  | hptot   0.0M  | hpuse   0.0M  |
SWP |  tot   980.0M |  free  179.0M |               |               |               |               |                |               |               |               | vmcom   4.4G  |               | vmlim   2.8G  |
PAG |  scan       0 |  steal      0 |               |  stall      0 |               |               |                |               |               |               | swin    1286  |               | swout      0  |
LVM |  1--vg-swap_1 |  busy     82% |               |  read    1286 |  write      0 |  KiB/r      4 |  KiB/w      0  |               | MBr/s    0.5  | MBw/s    0.0  | avq     1.02  |               | avio 6.31 ms  |
LVM |  BK1--vg-root |  busy      2% |               |  read       0 |  write     14 |  KiB/r      0 |  KiB/w      4  |               | MBr/s    0.0  | MBw/s    0.0  | avq     1.40  |               | avio 17.1 ms  |
DSK |   sda |  busy     84%     12% |               |  read    1253 |  write     10 |  KiB/r      4 |  KiB/w      6  |               | MBr/s    0.5  | MBw/s    0.0  | avq     1.03  |               | avio 6.46 ms  |
NFC |  rpc     4647 |  read 0 |              321.1M |  write      0 |  retxmit    0 |  autref  4647 |                | dwrit     0K  |               |               | mwrit     4K  |               |               |
NET |  transport    |  tcpi    4657 |  tcpo    4689 |  udpi 0 |  udpo 0 |  tcpao      0 |  tcppo      0  |               rit     0K  | tcprs      0  | tcpie      0  | tcpor      0  | udpnp      0  | udpie      0  |
NET |  network      |  ipi     4693 |               |  ipo     4689 |  ipfrw      0 |  deliv   4693 |                | dwrit     0K  |               |               | icmpi      0  |               | icmpo      0  |
NET |  enp0s25   0% |  pcki    4932 |  pcko    4690 |  sp 1000 Mbps |  si  894 Kbps |  so  705 Kbps |                | coll 0  | mlti|            5  | erri 0  | erro 0  | drpi  0K  |           80  | drpo       0  |

  PID SYSCPU USRCPU VGROW RGROW RDDSK WRDSK RUID EUID ST EXC THR S CPUNR CPU CMD 1/1|  tcpao     74 |  tcppo      8  |               | tcprs     30  | tcpie      0  | tcpor      7  | udpnp    130  | udpie      0  |
 1941 0.29s 0.7426     ipi    93322 |        0K 5440K 5195476 48899 |  i144w            4K root5252 |              root            --          - 1   |               D 0   i    171  |      12%         borg
  201 0.04s 0.00s60K 0K 0K 0K root7 |  pcko   50727 |  sp 1000 Mbps |  si 1319 Kbps |  so   43 Kbps |              root            -- - 1ti    1244  | erri          I            1 0% kworke1/1:1H
 1995 0.01s 0.01s18952K 660K 140K 0K rootk   0  319 |  sp    0 Mbps |  s  0  0 Kbps |  so    0 Kbps |              root            -- - 1ti       0  | erri          R            0 0% atop             atop
  925 0.01s 0.00s10K 0K 0K 0K r0ot root --   0                                      *** system and process activity since boot ***             -           3         S            1          0%         accounts-daemo
    8 0.01s 0.00s10K 0K 0K 0K r0ot root -- - 1 I 0 0% kworker/0:1H                                root             root                                    1         I            0                     rcu_sched
  199         0.01s         0.00s            0K             0K            0K            0K        root             root            --          -           1         I            0          0%         kworker/0:1H
 1991         0.00s         0.00s            0K             0K            0K            0K        root             root            --          -           1         I            1          0%         kworker/u16:3

```

**Comment by No15pecial** [comment]
OK So I built a new Ubuntu box (Ubuntu 16.04.6 LTS)  Dell Optiplex 755 4 GB memory. Gave it 20GB of swap. Moved my scripts and started the backup - We'll see if it was a SWAP Space issue.  Borg version is 1.0.12 (even lower that the Ubuntu 18 box which was Borg 1.1.5).

I just did a apt install borgbackup not sure how or what distro to use to get the latest and greatest

**Comment by ThomasWaldmann** [comment]
See there:

- https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/borgbackup  <- ubuntu .deb pkgs
- https://github.com/borgbackup/borg/releases  <- fat binary packages, use 1.1.x

**Comment by No15pecial** [comment]
Ok I will check those two links

Progress: after the new server with large swap Bork is still running but it usually does this for 3 days and then starts erroring every time so I'll comeback next week and update the progress 

Thanks for the support thus far

**Comment by No15pecial** [comment]
I stumbled across this ==> https://launchpad.net/~costamagnagianfranco/+archive/ubuntu/borgbackup <- ubuntu .deb pkgs but I didn't know who "Gianfranco Costamagna" is so I was reluctant to add the distro.

**Comment by No15pecial** [comment]
If I grab borg-linux64 from https://github.com/borgbackup/borg/releases <- fat binary packages, use 1.1.x where do I get borgfs from ?

I did update my distro on a test box with the following - Seems to work and show version 1.1.10

sudo add-apt-repository ppa:costamagnagianfranco/borgbackup
sudo apt-get update
apt install borgbackup

I did a wget of https://github.com/borgbackup/borg/releases/download/1.1.10/borg-linux64  made it executable and it reports to be v 1.1.10 but the sizes differ 

-rwxr-xr-- 1 root root 15007744 May 15 23:27 borg-linux64*

-rwxr-xr-x 1 root root 388 Sep  7 12:11 /usr/bin/borg*
lrwxrwxrwx 1 root root   4 Sep  7 12:11 /usr/bin/borgbackup -> borg*
-rwxr-xr-x 1 root root 392 Sep  7 12:11 /usr/bin/borgfs*

Any info to clear it up is appreciated.

**Comment by ThomasWaldmann** [comment]
Gianfranco Costamagna (aka @LocutusOfBorg ) is the borgbackup package maintainer of ubuntu and debian.

If you used the ubuntu or debian package, you already trusted him, so no problem using stuff from his ppa.

borgfs: never tried myself, but try symlinking borgfs -> borg.

The borg-linux64 download from github releases is a fat binary, much larger than the binary in the dist linux packages. This is because it contains everything needed to run (including python, libraries) on any recent 64bit linux system.

**Comment by No15pecial** [comment]
> Gianfranco Costamagna (aka @LocutusOfBorg ) is the borgbackup package maintainer of ubuntu and debian.
If you used the ubuntu or debian package, you already trusted him, so no problem using stuff from his ppa.
Good to know - I updated my PPA to Gianfranco on a test box but it sounds liek tiy are saying that Gianfranco's ppa comes embedded in ubuntu not sure why it install 1.0.12 out of the box. I had to 
```sudo add-apt-repository ppa:costamagnagianfranco/borgbackup
sudo apt-get update
sudo apt install borgbackup```
to get the newer version in the test box 
Knowing its good I'll do this on the new box I made as soon as the backup is done., Perhaps I'll blow away the  repo upgrade to 1.1.10  via above commands  and start again

**Comment by ThomasWaldmann** [comment]
This is normal debian/ubuntu packaging / distribution policy.

The **stable** dist is called stable, because usually versions don't change, but just get security fixes.
There can be exceptions (for good reasons), though.

Newer stuff you can usually get from `xxx-backports` source (debian) or from the maintainer's ppa (ubuntu).

**Comment by LocutusOfBorg** [comment]
@No15pecial hello from Debian/Ubuntu borg maintainer.
If you find a bug that is fixed in a newer borgbackup version, and you don't want to use my ppa (for good reasons, the archive should be fixed), please just open a bug report and subscribe me to it.
I'll take care of updating borgbackup even for stable.

Unfortunately stable updates needs to be carefully tested, to avoid risk of breaking stable releases.
This is why I maintain them in a ppa, but this doesn't stop me from updating it in case somebody asks for it and tests if everything is good.
I can't upload and test by myself, I prefer somebody other than me to check the package

**Comment by LocutusOfBorg** [comment]
Right now the Ubuntu package has zero bugs
https://bugs.launchpad.net/ubuntu/+source/borgbackup

the release team might ask me "if nobody complains why do you want to push an update with a risk of breaking the current working one"?

**Comment by No15pecial** [comment]
I hate to bother you and I did do a google search.  I think this should work but its not skipping the "snapshot "directories. 



borg create --verbose  --stats --chunker-params 19,23,21,4095 --show-rc  --exclude-caches --exclude '/NewCorpFiler-Mounts/GDC_IT/.snapshot'   --exclude '/NewCorpFiler-Mounts/GDC_ENG/.snapshot'  --exclude  '/NewCorpFiler-Mounts/Vol1/.snapshot'  --exclude  '/NewCorpFiler-Mounts/GDC_ENG_VM-BKS/.snapshot'    /Synology-Mounts/BorgHoard/GDCBorgRepo::BorgBK01-2019-12-12  /NewCorpFiler-Mounts/GDC_IT/ /NewCorpFiler-Mounts/GDC_ENG/  /NewCorpFiler-Mounts/Vol1/ /NewCorpFiler-Mounts/GDC_ENG_VM-BKS/  



    
 Best Regards,      
If you have any questions please do not hesitate to contact me     
----------------
 
           Domenic DiSorbo        |     Information Technology Manager    |         General DataComm, LLC.    
    DID: 203 266-2044    |     M: 203 560-6827    |             Domenic.DiSorbo@GDC.COM   |            WWW.GDC.COM   
     	        
                	        
           
            353 Christian ST. Suite 4           
           Oxford, CT. 06478-1053           
           WWW.GDC.COM           
           (203) 729-0271                  
 


 
----------------
 HIPAA/Confidential Notice- Please note:
 

 This Email along with all of the information in this e-mail message, including any attachments, is covered by the Electronic Communications Privacy Act, 18 U.S.C. §§ 2510-2521. This information is confidential, not for publication, and intended only for the sole use of the individual(s) named above. If you, the reader of this message, are not the intended recipient, you are hereby notified that you should not further disseminate, distribute, or forward this e-mail message. If you have received this e-mail message in error, please destroy all copies of the original message, and notify the sender.  GENERAL DATACOMM

**Comment by ThomasWaldmann** [comment]
@No15pecial do not put unrelated stuff to the end of random github issues.

https://borgbackup.readthedocs.io/en/stable/quickstart.html#automating-backups there is an example of using excludes. If that doesn't help, file a new issue please.

**Comment by ThomasWaldmann** [comment]
Is there anything left to do here?

**Comment by LocutusOfBorg** [comment]
No, I'll try to SRU borgbackup soon

**Comment by ThomasWaldmann** [comment]
@No15pecial did the more recent borg version fix the malloc issue?

**Comment by No15pecial** [comment]
> 
> 
> @No15pecial did the more recent borg version fix the malloc issue?

Its been so long , I can't really remember but I think my swap space was too small and that was causing the issue 

`OK So I built a new Ubuntu box (Ubuntu 16.04.6 LTS) Dell Optiplex 755 4 GB memory. Gave it 20GB of swap. Moved my scripts and started the backup - We'll see if it was a SWAP Space issue. Borg version is 1.0.12 (even lower that the Ubuntu 18 box which was Borg 1.1.5).`
